### DISCUSSION TITLE
IINA HDR off via MacBook m1 HDMI Denon AVR

### DISCUSSION THREAD

**Discussion by generaltest** [iina/iina#4578] [Q&A]
Hi team,

Plenty of people use AVRs to control sound and video. In this case I'm using Denon X4300h connected to Macbook Air M1 via HDMI. Then, AVR takes the video signal to projector. 
AVR and Projector are HDR compatible. 

If I want to mirror or extend the screen, macbook settings for display says "optimise for mac" or "optimise for Denon-AVRHD". If I choose the correct way which is mirror/extend the screen and "optimise for DENON", then the IINA HDR toggle is gray out and off. But if I switch to mirror/extend the screen and "optimise for macbook display" then HDR is ok. 

The impact is that if I optimise secondary display for DENON then I have full width of projected area (full width of secondary screen). But if it is optimised for macbook display then projected area is cut by 10%. Changing resolution on macobok doesn't make any difference. There is no setting on AVR or Projector that I can adjust.

Question/Request: can you make HDR enabled on all screens, not only macbook screen?

Thanks,
Greg

**Comment by low-batt** [comment]
IINA bases its decision on whether macOS indicates the screen supports [Extended Dynamic Range](https://developer.apple.com/documentation/metal/hdr_content/displaying_hdr_content_in_a_metal_layer#3335505).

In IINA's log look for messages that indicate the maxPotential EDR value:

```text
20:34:12.333 [iina][d] Refreshing HDR for player0 @ display1: "Built-in Retina Display" visible frame (0.0, 41.0, 1728.0, 1038.0) EDR: {supports=true maxPotential=16.0 maxCurrent=1.0}
```

From [Determining Support for EDR Values](https://developer.apple.com/documentation/metal/hdr_content/determining_support_for_edr_values#overview):
> To find out whether a display can support EDR values, read the [maximumPotentialExtendedDynamicRangeColorComponentValue](https://developer.apple.com/documentation/appkit/nsscreen/3180381-maximumpotentialextendeddynamicr) property on an [NSScreen](https://developer.apple.com/documentation/appkit/nsscreen) object for that display. A value greater than 1.0 indicates that the display supports EDR values; otherwise, the display supports only SDR values.

If macOS reports 1.0 for maxPotential then IINA decides the screen is incapable of supporting HDR.

My guess is that when you select "optimise for DENON" macOS reports EDR is not available. But that is just a guess. Check the IINA log see if that is what is happening.

From the Apple Support document [Play HDR video on Mac](https://support.apple.com/en-us/HT210980):
> Use an HDR10-compatible display that connects to the Thunderbolt  or HDMI port of your Mac.

From the DENON support document [HDR FORMAT COMPATIBILITY](https://support-eu.denon.com/app/answers/detail/a_id/4723/~/hdr-format-compatibility):
> Currently, all our **HDR (High Dynamic Range)** capable AVRs or soundbars support **HDR10** pass-through.

Seems like macOS should decide HDR is supported.

This is a nice small test that checks EDR capabilities: [macOS EDR Test](https://gist.github.com/Akemi/185d83afbef47c0c8e0add2a3c99b6d2)

What does the IINA log file show when IINA shows HDR is not available?

**Reply by generaltest** [reply]
Thank you for quick and detailed response.

I just connected my extra screen Dell P2217H directly via HDMI
If I mirror build in display and optimise for MacBook Air then HDR is available ON
This monitor is NOT HDR compatible yet toggle works
If I "optimise for Dell" then toggle is off
IINA settings > advance > I enabled Logging > restarted IINA
Log on mirror display & optimise for Mac:

10:06:20.594 [iina][d] Refreshing HDR for player0 @ display1: "Built-in Retina Display" visible frame (0.0, 85.0, 1440.0, 790.0) EDR: {supports=true maxPotential=2.0 maxCurrent=1.0}

Log on mirror display & optimise for Dell:

10:08:38.809 [iina][d] Refreshing HDR for player0 @ display2: "DELL P2217H" visible frame (0.0, 85.0, 1920.0, 970.0) EDR: {supports=false maxPotential=1.0 maxCurrent=1.0}

Im not a coder so not sure how to run SWIFT code.

But my conclusion is that this is not a problem with the device that is receiving the signal (external screen or DENON) but Mac or IINA. There are not many settings on Mac to play with so I doubt this can be solved on system level.

Last hope is IINA. Why would IINA listen to Mac for EDR response? Why IINA can't just let the user choose HDR ON or OFF? Clearly I can have HDR toggle ON on non HDR compatible screen if I optimise for Mac screen. So just let the user switch on/off the HDR and there will be problem with external screens having strange aspect ratio and HDR content?

Thanks

**Comment by low-batt** [comment]
What this test seems to tell us is that when mirroring macOS does not have the ability to offer a display profile that "works on both". The intersection between the capabilities of the two monitors. So you have to tell macOS which to use and the other monitor is treated as if that is ok. These test results are interesting, but this test is not testing the scenario we are concerned with, namely why does "optimise for DENON" disable HDR?

In the case we care about every device supports HDR, yes? The Mac, the DENON AVR, and the projector. And yet something about that setup causes macOS to decide the DENON does not support HDR. Possibly macOS is not understanding HDR pass-though or the DENON is not properly reflecting the HDR capabilities of the device attached to it.

On this question:
> Last hope is IINA. Why would IINA listen to Mac for EDR response? Why IINA can't just let the user choose HDR ON or OFF? 

Because IINA does not directly output to the displays. IINA must go through macOS. If macOS does not believe the display supports HDR then there is nothing IINA can do about that.

Don't worry about that separate test I mentioned. To run that might require installing development tools, which I do not recommend if you are not a developer. The IINA log file is giving us the critical info that we care about in this case, namely if macOS believes HDR is supported.

**Reply by generaltest** [reply]
I just checked Apple TV via Denon to Project and everything is in HDR10
I tried different input output ports and HDMI cables
It seems that macOS is not recognising Denon as HDR compatible (but Apple TV does) - weird .....

So I'm guessing this is dead end?
HDR output is controlled by Mac OS not IINA
If Mac OS do not recognise screen as HDR compatible then HDR won't go through and IINA will toggle it off
If so, I guess we can close this discussion.

Thank slow-batt for your support

**Comment by low-batt** [comment]
Indeed, this is an interesting result:
> I just checked Apple TV via Denon to Project and everything is in HDR10
I tried different input output ports and HDMI cables
It seems that macOS is not recognising Denon as HDR compatible (but Apple TV does) - weird 

That seems to indicate we should focus on the Mac and not the AVR.

The Apple Support document [Play HDR video on Mac](https://support.apple.com/en-us/HT210980) in the "Supported external HDR displays" section says: 
> _HDMI displays:_ Connect to the HDMI port. If your Mac doesn't have an HDMI port, connect to the Thunderbolt port using an HDMI cable and USB-C adapter such as the [Apple USB-C Digital AV Multiport Adapter](https://support.apple.com/kb/HT207806) or [Belkin USB-C to HDMI Adapter](https://www.apple.com/shop/product/HKQ22ZM/A/belkin-usb-c-to-hdmi-adapter).

And:
> If your Mac doesn't automatically enable HDR for the display, enable it in [Displays settings](https://support.apple.com/en-us/HT210980#checkmac).

Have you checked the macOS display settings to see if there is some sort of HDR setting when the AVR is connected?

I assume some sort of adapter is being used to attach the HDMI cable from the AVR to the Mac. Is it a no-name adapter or something trustworthy?

The DENON support document [NO HDR SIGNAL WITH APPLE TV 4K](https://support-eu.denon.com/app/answers/detail/a_id/6027/related/1) says:
> **Note:** The AVRs HDMI Diagnostics tool (if the AVR incorporates it) can be used to check if the cable is 18 Gbps.

If your AVR contains such a tool then possibly it could be used to confirm the HDMI signal from the MacBook Air is good.

There are differing views on what a GitHub issue is for. Many projects, and GitHub as well, treat them as a task list for developers. With that viewpoint, if the root cause of the problem is not in the code the project controls then the issue is closed.

My own opinion is that this is a developer centric view of the world which normal users are not expecting. If an application is malfunctioning that should be tracked somewhere. Doesn't matter to a user that the root cause is outside of code the project controls.

As IINA is built on top of a lot of libraries it is not uncommon for the root cause to be in a library IINA uses. We have also encountered defects in AppKit/macOS. IINA keeps these issues open and uses labels that start with `upstream:` to keep track of problems that have been tracked to code outside of the project.

Of course this is not an [issue](https://github.com/iina/iina/issues), it is a discussion. We don't track defects in discussions.

And of course when the root cause is outside of the IINA project there is only so much help an IINA developer can provide.

Before your response I poked around a bit and found [Denon AVR Not Passing HDR/Dolby Vision ](https://www.reddit.com/r/hometheater/comments/qmtbvl/denon_avr_not_passing_hdrdolby_vision/). As those posts are from 2 years ago they may no longer be applicable. They mention certain settings on the AVR that need to be enabled. The result from the Apple TV test seems to indicate the AVR is correctly configured, but I would check that thread to see if anything in it is still useful and double check the AVR settings are correct.
